lib/src/folding/foldable_block.dart

```diff
@@ -23,10 +25,44 @@ class FoldableBlock with EquatableMixin {
   bool includes(FoldableBlock other) {
     return startLine <= other.startLine && endLine >= other.endLine;
   }
+
+  bool intersects(FoldableBlock other) {
+    return other.endLine >= startLine &&
+        other.endLine < endLine &&
+        !includes(other);
+  }
+
+  FoldableBlock copy() {
+    return FoldableBlock(
+      startLine: startLine,
+      endLine: endLine,
+      type: type,
+    );
+  }
 }
 
 extension FoldableBlockList on List<FoldableBlock> {
   void sortByStartLine() {
     sort((a, b) => a.startLine - b.startLine);
   }
+
+  void joinIntersecting() {
+    if (length < 2) {
+      return;
+    }
+
+    for (int i = 1; i < length; i++) {
+      final currentBlock = this[i];
+      final previousBlock = this[i - 1];
+      if (currentBlock.intersects(previousBlock)) {
```

Comment:
`intersects(other)` is not needed at all. Just check if `previous.endLine >= current.startLine && previous.endLine < current.endLine` or maybe make a method that encapsulates this (but I am struggling to come up with a name, maybe `intersectsWithEnd(int)`)

Reply:
This logic falls on this test:
```
        _Example(
          'Python. Nested highlight blocks',
          code: '''
a = [[
  5
  ]
]''',
          expected: [
            _FB(startLine: 0, endLine: 2, type: _T.brackets),
            _FB(startLine: 0, endLine: 3, type: _T.brackets),
          ],
        ),
```

Reply:
Let's disallow two blocks to start on the same line because we can only have one folding button per line. Only the outer block of the two should survive.